src/App.tsx

```diff
@@ -27,153 +23,157 @@ export type CountMetadata = {
+    history.push({ deck, city, index, destroy: true })
+    setLastUndo('destroy');
+    if (deck === 'deck') {
+      const newKnownDeck = [ ...knownDeck ];
+      newKnownDeck[index] = incrCity(newKnownDeck[index], city, -1);
+      setKnownDeck(newKnownDeck)
+    } else {
+      setDiscard(incrCity(discard, city, -1));
+    }
   };
 
-  const handleDraw = (oldCount: DeckCount, setOldCount: (arg: DeckCount) => void, city: string) => {
-    const newCounts = _.cloneDeep(oldCount);
-    newCounts[city].count = oldCount[city].count - 1;
-    const card = cityToCard(city);
-    setOldCount(newCounts);
-    setDiscard(discard.concat([card]));
+  const handleDraw = (city: string, index: number) => {
+    history.push({
+      deck: 'deck',
+      city,
+      index,
+      destroy: false,
+    });
+    setLastUndo('draw');
+    const newKnownDeck = [...knownDeck];
+    newKnownDeck[index] = incrCity(newKnownDeck[index], city, -1);
+    if (Object.keys(newKnownDeck[index]).length === 0) {
+      newKnownDeck.splice(index, 1);
+    }
+    setKnownDeck(newKnownDeck);
+    setDiscard(incrCity(discard, city));
   };
 
   // handleUndoDraw takes the most recently-drawn card out of the discard and re-adds it to the set passed in the parameter
-  const handleUndoDraw = (oldCount: DeckCount, setOldCount: (arg: DeckCount) => void) => {
-    const newCounts = _.cloneDeep(oldCount);
- 
-    const undoCard = discard[discard.length - 1];
-    const city = undoCard.city;
-
-    newCounts[city].count = oldCount[city].count + 1;
-    const card = cityToCard(city);
-    setOldCount(newCounts);
-    setDiscard(discard.slice(0, discard.length - 1));
-  }
+  const handleUndoDraw = () => {
+    const draw = history.pop();
+    if (!draw) return;
+
+    const prevUndo = history[history.length - 1];
+    if (!prevUndo) {
+      setLastUndo(null);
+    } else if (prevUndo === 'shuffle') {
+      setLastUndo('shuffle');
+    } else if (prevUndo.destroy) {
+      setLastUndo('destroy');
+    } else {
+      setLastUndo('draw');
+    }
 
-  // Shuffles the discard pile to the top of the deck, updating counts in the known deck if appropriate
-  const handleShuffleDiscard = () => {
-    const newKnownDeck = _.cloneDeep(knownDeck);
-    for (let card of discard) {
-      const {city} = card;
-      if (newKnownDeck.hasOwnProperty(card.city)) {
-        newKnownDeck[city].count = newKnownDeck[city].count + 1;
-      } else {
-        newKnownDeck[city] = {city, count: 1};
+    if (draw === 'shuffle') {
+      const newKnownDeck = [...knownDeck];
+      setDiscard(newKnownDeck.pop()!);
+      setKnownDeck(newKnownDeck);
+      return;
+    }
+
+    const { city, index, deck, destroy } = draw;
+
+    if (deck === 'deck') {
+      const newKnownDeck = [...knownDeck];
+      if (index >= knownDeck.length) {
+        newKnownDeck.push({})
       }
+      newKnownDeck[index] = incrCity(newKnownDeck[index], city);
```

Comment:
You're right, there was a bug here. I needed to keep empty subdecks that are in the middle of the deck otherwise the index tracking gets all messed up. This new patch only pops from the deck when the topmost subdeck runs out of cards. This keeps the indices in the history consistent.